src/Jackalope/Version/VersionHandler.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+namespace Jackalope\Version;
+
+use Jackalope\Node;
+use Jackalope\Transport\TransportInterface;
+use PHPCR\UnsupportedRepositoryOperationException;
+
+class VersionHandler
+{
+    const MIX_VERSIONABLE = 'mix:versionable';
+    const MIX_SIMPLE_VERSIONABLE = 'mix:simpleVersionable';
+
+    /**
+     * @var TransportInterface
+     */
+    private $transport;
+
+    public function __construct(TransportInterface $transport)
+    {
+        $this->transport = $transport;
+    }
+
+    public static function addVersionProperties(Node $node, TransportInterface $transport)
+    {
+        // TODO solve without reflection
+        $reflection = new \ReflectionMethod(get_class($transport), 'storeNode');
+        $reflection->setAccessible(true);
+
+        // TODO move this to the version handler
+        $session = $node->getSession();
+        $rootNode = $session->getRootNode();
+
+        $node->setProperty('jcr:isCheckedOut', true);
+
+        if (!$rootNode->hasNode('jcr:system/jcr:versionStorage')) {
+            if (!$rootNode->hasNode('jcr:system')) {
+                $rootNode->addNode('jcr:system');
+            }
+            $rootNode->addNode('jcr:system/jcr:versionStorage');
+        }
+
+        $versionStorageNode = $rootNode->getNode('jcr:system/jcr:versionStorage');
+        if (!$versionStorageNode->hasNode($node->getIdentifier())) {
+            $versionStorageNode->addNode($node->getIdentifier(), 'nt:versionHistory');
+        }
+        $versionHistory = $versionStorageNode->getNode($node->getIdentifier());
+        $versionHistory->setProperty('jcr:versionableUuid', $node->getIdentifier());
+
+        // TODO Set jcr:copiedFrom if needed
+
+        $versionHistory->addNode('jcr:versionLabels', 'nt:versionLabels');
+        $rootVersion = $versionHistory->addNode('jcr:rootVersion', 'nt:version');
+        $reflection->invoke($transport, $rootVersion->getPath(), $rootVersion->getProperties());
```

Comment:
That's horrible, but I couldn't find a way around using reflections...

The problem here is that the Nodes have to be persisted before they can be referenced. That's why I call here (and a few lines later) the protected `storeNode` method via a reflection. That's only necessary because there is no way to save a single node. If `$session->save()` is called the system ends up in an inconsistent state (because some nodes are being created twice) and a rollback is executed.

How should we handle this? /cc @dbu